app/controllers/admin/api/applications_controller.rb

```diff
@@ -24,13 +24,7 @@ def application_filter_params
   end
 
   def applications
-    @applications ||= begin
-      cinstances = current_account.provided_cinstances.where(service: accessible_services)
-      if (service_id = params[:service_id])
-        cinstances = cinstances.where(service_id: service_id)
```

Comment:
This is already being done by `scope_search` in `#index`. I was thinking whether this `scope_serach` call should be moved here inside `#applications` because `#application` also calls `#applications`.

But I think it doesn't make sense for `#application` because there a single app is selected by other parameters.

Reply:
This is the actual fix for the query in JIRA, right? The rest are just N+1 improvements?

Reply:
`provided_by` is the main thing. This is just further optimization.

Reply:
> But I think it doesn't make sense for `#application` because there a single app is selected by other parameters.

I actually don't agree with this.

We do have `service_id` as a filtering parameter in the API docs. And we actually need to conserve it, because we allow applications with the same app ID under different services. So, we should be able to find the application that belongs to the specified service.

![application-find](https://github.com/user-attachments/assets/1fa9d530-2558-4acd-8fcd-b515aa8e0e4d)


I tried these three tests (in `test/integration/user-management-api/applications_test.rb`). They all pass in `master` but the 2nd and the 3rd fail in the branch.

```
  test 'find by user_key with correct service_id' do
    @service.backend_version = '1'
    @service.save!

    get find_admin_api_applications_path(format: :xml), params: { user_key: @application.user_key,
                                                                   provider_key: @provider.api_key,
                                                                   service_id: @service.id }

    assert_response :success
    assert_application(@response.body,
                       { id: @application.id,
                         user_key: @application.user_key })
  end

  test 'find by user_key with incorrect service_id' do
    @service.backend_version = '1'
    @service.save!

    another_service = FactoryBot.create(:service, account: @provider)

    get find_admin_api_applications_path(format: :xml), params: { user_key: @application.user_key,
                                                                   provider_key: @provider.api_key,
                                                                   service_id: another_service.id }

    assert_response :not_found
  end

  test 'find by user_key when there are two apps with the same user key' do
    @service.backend_version = '1'
    @service.save!

    another_service = FactoryBot.create(:service, account: @provider, backend_version: '1')
    another_app = FactoryBot.create(:cinstance,
                                    service: another_service,
                                    plan: FactoryBot.create(:application_plan, issuer: another_service),
                                    user_key: @application.user_key)

    get find_admin_api_applications_path(format: :xml), params: { user_key: @application.user_key,
                                                                  provider_key: @provider.api_key,
                                                                  service_id: another_service.id }

    assert_application(@response.body,
                       { id: another_app.id,
                         service_id: another_service.id,
                         user_key: another_app.user_key })
  end
```

Reply:
Thanks a lot for catching this!

I added your tests and updated controller to add the service condition when searching by `user_key`. See https://github.com/3scale/porta/pull/3845/commits/fc5f1e727ed12458eec6a9f2aef353232e9371dd . Is this enough? `app_id` should always be unique as far as I understand so it doesn't need such restriction. Correct?

Reply:
Nope, it's actually the same for `app_id` :) we allow having the same ID (and the app key too) as long as the apps are under different services.

Reply:
:face_with_spiral_eyes: ok
```
      scope.where.has { (service.backend_version == '1') & (user_key == param_key) }.first!

                   when app_id = params[:app_id]
      scope.where.has { (service.backend_version != '1') & (application_id == app_id) }.first!

                     else
      scope.find(params[:application_id] || params[:id])
```

So it sounds like `application_id` and `app_id` are not the same. How about `application_id`, do we need it to be `scope.find(params[:application_id] || params[:id])` or `applications.find(params[:application_id] || params[:id])
` is sufficient (in case this finally is the primary key and can't be anything else)?

Reply:
Yes, right, `application_id` seems to the be `id` of the Cinstance (aka primary key). It seems a bit weird using it as a search query parameter... But well. To answer your question I'd say we need to check the current behavior. To me it seems logical to apply `find` to the already filtered scope. So, if you have the following:

```
service: 1
  application: 
    id: 11
    app_id: a1
    
service: 2
  application: 
    id: 22
    app_id: a2
```

searching by `application_id=11&app_id=a1&service_id=1` should return the first application, while any other combination should return a 404, e.g.:
`application_id=22&app_id=a2&service_id=1`
`application_id=22&app_id=a1&service_id=2`

Reply:
If you have `app_id` then `application_id` appears to be ignored. Historically using the wrong service_id would still return nothing though.

Reply:
Updated, PTAL